### PR TITLE
✨ Add in memory provider

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
Add the in-memory provider to be used for stress testing

NOTE: keeping the commits separated to allow faster review

### CODE REVIEW

**Comment by killianmuldoon** [comment]
Link checker issue is fine as it's just a link to a folder introduced in this PR.

File: test/infrastructure/inmemory/templates/cluster-template-inmemory-development.yaml [hunk 1]
```diff
@@ -0,0 +1 @@
+---
```
Comment (sbueringer) [inline test/infrastructure/inmemory/templates/cluster-template-inmemory-development.yaml:1]: As we are publishing this file on our release. let's have a follow-up to add an actual working cluster-template here

Comment (sbueringer) [reply test/infrastructure/inmemory/templates/cluster-template-inmemory-development.yaml:1]: I wonder if we should do some sort of alpha release to verify this all works

Comment (fabriziopandini) [reply test/infrastructure/inmemory/templates/cluster-template-inmemory-development.yaml:1]: I have added the template I'm using for experimenting with the new provider. 
I will open a follow-up issue to improve it

Comment (sbueringer) [reply test/infrastructure/inmemory/templates/cluster-template-inmemory-development.yaml:1]: I think @ykakarap can update it when he's working on the ClusterClass support (with a template using a CC)

**Comment by fabriziopandini** [comment]
Note:
- People can still review and we'll open follow-up PRs to address potential findings. 
- This work is still at the initial stage - it is just a proof of concept ATM -; so another way to provide feedback is to follow upcoming PRs in this area (I will open issues tracking the next steps and link it here) 
- I'm also planning to propose to the community a deep dive session where I can give an overview of the internals of the new provider to everyone interested in it.

**Pull request merged**